src/sdam/topology.ts

```diff
@@ -411,17 +402,26 @@ export class Topology extends TypedEventEmitter<TopologyEvents> {
   }
 
   /** Initiate server connect */
-  connect(callback: Callback): void;
-  connect(options: ConnectOptions, callback: Callback): void;
-  connect(options?: ConnectOptions | Callback, callback?: Callback): void {
-    if (typeof options === 'function') (callback = options), (options = {});
+  async connect(options?: ConnectOptions): Promise<Topology> {
+    if (this.connectionLock) {
+      return this.connectionLock;
+    }
+
+    try {
+      this.connectionLock = this._connect(options);
+      await this.connectionLock;
+    } finally {
+      // release
+      this.connectionLock = undefined;
+    }
```

Comment:
(optional) thoughts?

```suggestion
    this.connectionLock ??= this._connect(options);
    try {
    	await this.connectionLock;
    	return this;
    } finally {
    	this.connectionLock = undefined;
    }
```

Reply:
```suggestion
    try {
      return await (this.connectionLock ??= this._connect(options));
    } finally {
      this.connectionLock = undefined;
    }
```

Let's play ⛳ (kidding... mostly 👀)

Reply:
```suggestion
    return await (this.connectionLock ??= this._connect(options)).finally(() => this.connectionLock = undefined);
```

😤